arch/xtensa/src/esp32s2/esp32s2_i2c.c

```diff
@@ -0,0 +1,1702 @@
+          else /* Write operation */
+            {
+              i2c_traceevent(priv, I2CEVENT_SENDBYTE, priv->bytes,
+                             getreg32(I2C_SR_REG(priv->id)));
+              i2c_senddata(priv);
+
+              /* Finally sent the entire message. Update the I2C state to
+               * send a stop signal in the next interrupt.
+               */
+
+              if (priv->bytes == msg->length)
+                {
+                  if ((msg->flags & I2C_M_NOSTOP) == 0)
+                    {
+                      priv->i2cstate = I2CSTATE_STOP;
+                    }
+#ifndef CONFIG_I2C_POLLED
+                  else
+                    {
+                      priv->i2cstate = I2CSTATE_FINISH;
+                    }
+#endif
+                }
+            }
+        }
+      else if (priv->i2cstate == I2CSTATE_STOP)
+        {
+          /* Transmitted all data. Finish the transaction sending a stop
+           * and update the I2C state.
+           */
+
+          i2c_traceevent(priv, I2CEVENT_STOP, msg->length,
+                         getreg32(I2C_SR_REG(priv->id)));
+          i2c_sendstop(priv);
+#ifndef CONFIG_I2C_POLLED
+          priv->i2cstate = I2CSTATE_FINISH;
+#endif
+        }
+#ifndef CONFIG_I2C_POLLED
+      else if (priv->i2cstate == I2CSTATE_FINISH)
+        {
+          /* Disable all interrupts and release the semaphore */
+
+          i2c_intr_disable(priv);
+          nxsem_post(&priv->sem_isr);
+        }
+#endif
+    }
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp32s2_i2cbus_initialize
+ *
+ * Description:
+ *   Initialize the selected I2C port. And return a pointer to an unique
+ *   instance of struct i2c_master_s. This function may be called to obtain
+ *   multiple instances of the interface, each of which may be set up with a
+ *   different frequency and slave address.
+ *
+ * Parameters:
+ *   port          - Port number of the I2C interface to be initialized.
+ *
+ * Returned Value:
+ *   Pointer to valid I2C device structure is returned on success.
+ *   A NULL pointer is returned on failure.
+ *
+ ****************************************************************************/
+
+struct i2c_master_s *esp32s2_i2cbus_initialize(int port)
+{
+  irqstate_t flags;
+  struct esp32s2_i2c_priv_s *priv;
+#ifndef CONFIG_I2C_POLLED
+  const struct esp32s2_i2c_config_s *config;
+  int ret;
+#endif
+
+  switch (port)
+    {
+#ifdef CONFIG_ESP32S2_I2C0
+    case ESP32S2_I2C0:
+      priv = &g_esp32s2_i2c0_priv;
+      break;
+#endif
+#ifdef CONFIG_ESP32S2_I2C1
+    case ESP32S2_I2C1:
+      priv = &g_esp32s2_i2c1_priv;
+      break;
+#endif
+    default:
+      return NULL;
+    }
+
+  flags = enter_critical_section();
+
+  if ((volatile int)priv->refs++ != 0)
```

Comment:
I'm not sure why `volatile int` cast is needed here. Maybe better if `refs` itself is a `volatile int` and drop the cast here